hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java

```diff
@@ -1216,37 +1222,50 @@ public FileStatus getFileStatus(Path f) throws IOException {
           theInternalDir.getChildren().entrySet()) {
         INode<FileSystem> inode = iEntry.getValue();
+        Path path = new Path(inode.fullPath).makeQualified(myUri, null);
         if (inode.isLink()) {
           INodeLink<FileSystem> link = (INodeLink<FileSystem>) inode;
+
+          if (showMountLinksAsSymlinks) {
+            // To maintain backward compatibility, with default option(showing
+            // mount links as symlinks), we will represent target link as
+            // symlink and rest other properties are belongs to mount link only.
+            result[i++] =
+                new FileStatus(0, false, 0, 0, creationTime, creationTime,
+                    PERMISSION_555, ugi.getShortUserName(),
+                    ugi.getPrimaryGroupName(), link.getTargetLink(),
+                    path);
+            continue;
+          }
+
+          //  We will represent as non-symlinks. Here it will show target
+          //  directory/file properties like permissions, isDirectory etc on
+          //  mount path. The path will be a mount link path and isDirectory is
+          //  true if target is dir, otherwise false.
+          String linkedPath = link.getTargetFileSystem().getUri().getPath();
+          if ("".equals(linkedPath)) {
+            linkedPath = "/";
+          }
           try {
-            String linkedPath = link.getTargetFileSystem().getUri().getPath();
-            if("".equals(linkedPath)) {
-              linkedPath = "/";
-            }
             FileStatus status =
                 ((ChRootedFileSystem)link.getTargetFileSystem())
                 .getMyFs().getFileStatus(new Path(linkedPath));
-            result[i++] = new FileStatus(status.getLen(), false,
-              status.getReplication(), status.getBlockSize(),
-              status.getModificationTime(), status.getAccessTime(),
-              status.getPermission(), status.getOwner(), status.getGroup(),
-              link.getTargetLink(),
-              new Path(inode.fullPath).makeQualified(
-                  myUri, null));
+            result[i++] = new FileStatus(status.getLen(), status.isDirectory(),
+                status.getReplication(), status.getBlockSize(),
+                status.getModificationTime(), status.getAccessTime(),
+                status.getPermission(), status.getOwner(), status.getGroup(),
+                null, path);
           } catch (FileNotFoundException ex) {
-            result[i++] = new FileStatus(0, false, 0, 0,
-              creationTime, creationTime, PERMISSION_555,
-              ugi.getShortUserName(), ugi.getPrimaryGroupName(),
-              link.getTargetLink(),
-              new Path(inode.fullPath).makeQualified(
-                  myUri, null));
+            LOG.warn("Cannot get one of the children's(" + path
+                + ")  target path(" + link.getTargetFileSystem().getUri()
+                + ") file status.", ex);
+            throw ex;
```

Comment:
Should we just fallback to showing link permissions if the directory isn't found? as it was before.
Ls command throwing FNF for a child seems little weird. If the child isn't there the Ls need not to list it. FNF from Ls should be only I guess when the directory on which we are calling getListing() isn't available, For a child if it isn't present in the destination we can just log it and revert back to previous behavior and we can document that too as well, if the actual destination is present then only it shows destination permissions.

Reply:
This is a non-symlink flow: what we should assume for isDirectory when target not available?
I feel that users should know that target is not available somehow. In symlink case I am fine as we don't need that isDir info and anyway we will just keep target in symlink.
If we simply return successful when one of the child not not available, they should not assume is exist would be successful. 
Or we will simply return is directory as true when target not found and rest all permission are fake. Just logging does not help here.
Do you know what happens in router? probably we can make it to similar behavior for consistency in this case?

Reply:
In Router if the destination isn't present it will not throw exception, it will show the permissions and owner of the mount entry. In router the mount table entries also have owner/permissions. It will first try to get it from the actual destination, if the actual destination is not available then from the mount table. if Unable to get from Mount Entry(this won't happen general) then it will show owner as super user
isDir stays true always for mount entries assuming only directories will be mounted in general case.

Reply:
The problem here is, rounter has state, but in viewFS is simply light, state can be build from config only, so no real permissions notion. All mounts will have same perms bit. Just showing that really may not help. Most of the fs what we mount are high available. SO, don't  expect this case should trigger ( even I see this point in you comment "this won't happen general"). Even if it triggered, application will anyway starts seeing failures on further ops on that directory. And one more point I am having is, restricting first and relaxing later is safer than relaxing and later restricting is tough. So, I feel Let's throw exception. If this really turning out problems, then we can always relax (that should be safe as we are having general IOException). When soming to symlink notion, you always have an identity other than file/dir, that is link. So, I have no issues. But non-symlinks case, I am not really show some temp values there. With that values anyway app is going to fail if he crawls one more level.